app_dart/lib/src/request_handlers/update_branches.dart

```diff
@@ -0,0 +1,124 @@
+    this.processRunner,
+  }) : super(config: config);
+
+  final DatastoreServiceProvider datastoreProvider;
+  ProcessRunner? processRunner;
+
+  static const int kActiveBranchActivityPeriod = 7;
+
+  @override
+  Future<Body> get() async {
+    final DatastoreService datastore = datastoreProvider(config.db);
+    final KeyHelper keyHelper = config.keyHelper;
+
+    await _updateBranchesForAllRepos(config, datastore);
+
+    final List<BranchWrapper> branches = await datastore
+        .queryBranches()
+        .where((md.Branch b) =>
+            DateTime.now().millisecondsSinceEpoch - b.lastActivity! <
+            const Duration(days: kActiveBranchActivityPeriod).inMilliseconds)
+        .map<BranchWrapper>((md.Branch branch) => BranchWrapper(branch, keyHelper.encode(branch.key)))
+        .toList();
+    return Body.forJson(branches);
+  }
+
+  Future<void> _updateBranchesForAllRepos(Config config, DatastoreService datastore) async {
+    DateTime timeNow = DateTime.now();
+
+    processRunner ??= ProcessRunner();
+    final Set<RepositorySlug> slugs = Config.supportedRepos;
+    for (RepositorySlug slug in slugs) {
+      ProcessRunnerResult result =
+          await processRunner!.runProcess(['git', 'ls-remote', '--heads', 'git@github.com:flutter/${slug.name}']);
+      List<String> shaAndName = result.stdout.trim().split(RegExp(' |\t|\r?\n'));
+      List<String> branchShas = [];
+      List<String> branchNames = [];
+      for (int i = 0; i < shaAndName.length; i += 2) {
+        branchShas.add(shaAndName[i]);
+        branchNames.add(shaAndName[i + 1].replaceAll('refs/heads/', ''));
+      }
+      final GitHub github = await config.createGitHubClient(slug: slug);
+      await _updateBranchesForRepo(branchShas, branchNames, github, slug, datastore, timeNow);
+    }
+    return;
+  }
+
+  Future<void> _updateBranchesForRepo(List<String> branchShas, List<String> branchNames, GitHub github,
+      RepositorySlug slug, DatastoreService datastore, DateTime timeNow) async {
+    List<md.Branch> updatedBranches = [];
+    for (int i = 0; i < branchShas.length; i += 1) {
+      final RepositoryCommit branchCommit = await github.repositories.getCommit(slug, branchShas[i]);
+      int lastUpdate = branchCommit.commit!.committer!.date!.millisecondsSinceEpoch;
+      if (lastUpdate > timeNow.subtract(const Duration(days: kActiveBranchActivityPeriod)).millisecondsSinceEpoch) {
+        final String id = '${slug.fullName}/${branchNames[i]}';
+        final Key<String> key = datastore.db.emptyKey.append<String>(Branch, id: id);
+        updatedBranches.add(md.Branch(key: key, lastActivity: lastUpdate));
+      }
+    }
+    await datastore.insert(updatedBranches);
+  }
+}
+
+// FOR REVIEW: Probably related to how keyhelper is implementated,
+// but I have to add a wrapper class to avoid `Converting object to an encodable object failed: Instance of 'SerializableBranch'` error
+class BranchWrapper {
```

Comment:
Why do we have BranchWrapper and SerializableBranch? Wrapper and serializable are doing the same function

Reply:
umm I spent another hour looking at this, and guess that I probably need some help on understanding the internals. the closest example from getStatus was that it uses a `keyHelper.encode(status.commit.key)` , and had a commitStatus class wrapped around the commit class which has an internal key. If I remove JsonKey or keyHelper or wrapper class in code, my variables become empty or  json.decode would fail. The only way I can make current code run is to do an exact copy of getstatus (copying both keyhelper to help with the underlying key in branch class, and the wrapper class). I think you are right that there should be a better way to do this, but I haven't figure that out yet

Reply:
Skip keyhelper, and directly use `key.id`